kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/AdaptiveBigQueryWriter.java

```diff
@@ -74,17 +80,27 @@ public AdaptiveBigQueryWriter(BigQuery bigQuery,
   public void performWriteRequest(PartitionedTableId tableId,
                                   List<InsertAllRequest.RowToInsert> rows,
                                   String topic) {
-    InsertAllRequest request = createInsertAllRequest(tableId, rows);
-    InsertAllResponse writeResponse = bigQuery.insertAll(request);
-    // Should only perform one schema update attempt; may have to continue insert attempts due to
-    // BigQuery schema updates taking up to two minutes to take effect
-    if (writeResponse.hasErrors()
-        && onlyContainsInvalidSchemaErrors(writeResponse.getInsertErrors())) {
-      try {
-        schemaManager.updateSchema(tableId.getBaseTableId(), topic);
-      } catch (BigQueryException exception) {
-        throw new BigQueryConnectException("Failed to update table schema for: " + tableId.getBaseTableId(), exception);
+    InsertAllResponse writeResponse = null;
+    InsertAllRequest request = null;
+
+    try {
+      request = createInsertAllRequest(tableId, rows);
+      writeResponse = bigQuery.insertAll(request);
+      // Should only perform one schema update attempt; may have to continue insert attempts due to
+      // BigQuery schema updates taking up to two minutes to take effect
+      if (writeResponse.hasErrors()
+              && onlyContainsInvalidSchemaErrors(writeResponse.getInsertErrors())) {
+        attemptSchemaUpdate(tableId, topic);
+      }
+    } catch (BigQueryException e) {
+      if (isTableMissingSchema(e)) {
```

Comment:
if I understand correctly, there are two possiblities where we need to update the table schema:

1. the table exists and has a schema, but doesn't have some new cols (and thus needs to be updated)
2. the table exists but has no/an empty schema. This is new because this is only possible with the MemorySchemaRetriever.

is my understanding accurate?

Reply:
Yes, that's correct